zaza/utilities/run_report.py

```diff
@@ -0,0 +1,196 @@
+    if not run_data or reset:
+        run_data = {
+            ReportKeys.EVENTS: {},
+            ReportKeys.METADATA: {}}
+
+
+def register_event(event_name, event_state, timestamp=None):
+    """Register that event_name is at event_state.
+
+    :param event_name: Name of event
+    :type event_name: str
+    :param event_state: Name of event state (should be one of EventStates).
+    :type event_state: enum.Enum
+    :param timestamp: Seconds since epoch when event_state of event_name was
+                      reached.
+    :type timestamp: float
+    """
+    global run_data
+    _init_run_data()
+    timestamp = timestamp or time.time()
+    events = run_data[ReportKeys.EVENTS]
+    if event_name in events:
+        events[event_name][event_state] = timestamp
+    else:
+        events[event_name] = {event_state: timestamp}
+
+
+def register_metadata(cloud_name=None, model_name=None, target_bundle=None):
+    """Add metadata about this run.
+
+    :param cloud_name: Name of cloud eg canonistack1
+    :type cloud_name: str
+    :param model_name: Name of model
+    :type model_name: str
+    :param target_bundle: Name of bundle
+    :type target_bundle: str
+    """
+    global run_data
+    _init_run_data()
+    if cloud_name:
+        run_data[ReportKeys.METADATA]['cloud_name'] = cloud_name
+    if model_name:
+        run_data[ReportKeys.METADATA]['model_name'] = model_name
+    if target_bundle:
+        run_data[ReportKeys.METADATA]['target_bundle'] = target_bundle
+
+
+def get_events_start_stop_time(events):
+    """Return the time of the first event and the last.
+
+    :param events: Dict of events.
+    :type events: dict
+    :returns: Tuple of start time and endtime in seconds since epoch.
+    :rtype: (float, float)
+    """
+    start_time = -1
+    finish_time = -1
+    for event_name, event_info in events.items():
+        for event_state, timestamp in event_info.items():
+            if start_time == -1 or start_time > timestamp:
+                start_time = timestamp
+            if finish_time == -1 or finish_time < timestamp:
+                finish_time = timestamp
+    return start_time, finish_time
+
+
+def get_event_report():
+    """Produce report based on current run.
+
+    :returns: Dictionary of run data with summary info.
+    :rtype: dict
+    """
+    global run_data
+    _init_run_data()
+    report = copy.deepcopy(run_data)
+    start_time, finish_time = get_events_start_stop_time(get_events())
+    full_run_time = finish_time - start_time
+    for name, info in report[ReportKeys.EVENTS].items():
+        try:
+            event_time = (info[EventStates.FINISH] -
+                          info[EventStates.START])
+            info[ReportKeys.ELAPSED_TIME] = event_time
+            info[ReportKeys.PCT_OF_RUNTIME] = math.ceil(
+                (event_time / full_run_time) * 100)
+        except KeyError:
+            pass
+    return report
+
+
+def write_event_report(output_file=None):
```

Comment:
This is really picky, but this is really three functions.  One to generate the yaml report, one to log it, and one to write it to a file.  By conflating them, it makes it harder (fractionally) to test them as you have to mock out logging and `open` to test the yaml dumping, etc.  Do you see what I mean?